Zend/Optimizer/zend_inference.c

```diff
@@ -74,12 +74,34 @@
 	} \
 } while (0)
 
+#ifdef SYM_RANGE
+# define STORE_CURRENT_SYM recursion_stack[recursion_stack_idx].current_sym = sym_p
+#else
+# define STORE_CURRENT_SYM
+#endif
+
 #define CHECK_SCC_VAR(var2) \
 	do { \
 		if (!ssa->vars[var2].no_val) { \
 			if (dfs[var2] < 0) { \
-				zend_ssa_check_scc_var(op_array, ssa, var2, index, dfs, root, stack); \
+				/* Need to come back from "recursion" */ \
+				recursion_stack[recursion_stack_idx].var = var; \
+				recursion_stack[recursion_stack_idx].current_use = use; \
+				recursion_stack[recursion_stack_idx].current_phi = p; \
+				STORE_CURRENT_SYM; \
+				recursion_stack_idx++; \
+				/* "Recurse" on next var */ \
+				var = var2; \
+				use = ssa->vars[var].use_chain; \
+				p = ssa->vars[var].phi_use_chain; \
+				sym_p = ssa->vars[var].sym_use_chain; \
+				dfs[var] = *index; \
+				(*index)++; \
+				root[var] = var; \
+				goto recurse; \
```

Comment:
Maintaining of stack with 3 additional lists looks a bit weird to me.
Did you use some reference algorithm implementation or invent this yourself?